Add the support for generating sharedPreferencesForWebProcess which takes connection

rniwa opened the pull request:
#### 38e36fd479a284a1597097630ad87b2bc7bb9396
<pre>
Add the support for generating sharedPreferencesForWebProcess which takes connection
<a href="https://bugs.webkit.org/show_bug.cgi?id=277475">https://bugs.webkit.org/show_bug.cgi?id=277475</a>

Reviewed by Chris Dumez.

This PR adds the support for SharedPreferencesNeedsConnection extended attribute on
an IPC message receiver which, if specified on a message receiver, will generate
sharedPreferencesForWebProcess calls with connection as an argument. Also changed
the delimiter for specifying multiple feature flags from &quot;,&quot; to &quot;&amp;&amp;&quot;.

In addition, this PR consolidates multiple calls to sharedPreferencesForWebProcess
in the generated didReceiveMessage function to one call at the beginning of
the function to avoid the overhead of calling the function multiple times.

Finally, use this new capability in WebProcessPool.messages.in.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(generate_enabled_by):
(generate_runtime_enablement):
(async_message_statement):
(sync_message_statement):
(generate_enabled_by_for_receiver):
(generate_message_handler):
* Source/WebKit/Scripts/webkit/model.py:
(MessageReceiver.__init__):
(generate_global_model):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::messageArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledBy.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
(WebKit::TestWithEnabledBy::didReceiveMessage):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithEnabledBy_MultiConditionallyEnabled&gt;):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessages.h:
(Messages::TestWithEnabledBy::MultiConditionallyEnabled::name):
(Messages::TestWithEnabledBy::MultiConditionallyEnabled::arguments):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessageReceiver.cpp:
(WebKit::TestWithEnabledIf::didReceiveMessage):
* Source/WebKit/UIProcess/WebProcessPool.messages.in:

Canonical link: <a href="https://commits.webkit.org/281977@main">https://commits.webkit.org/281977@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4aeeab05fc8e6f9357844a8b1e291dfee9d9dbb2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40995 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14233 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65616 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12187 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48680 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12458 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49726 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8462 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64711 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38091 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53395 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30559 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/61152 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/34752 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11118 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56555 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10925 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67344 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5585 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57105 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5610 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53344 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57325 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4591 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9282 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37880 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/38976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/37625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/UIProcess/WebProcessPool.messages.in:
```diff
@@ -20,16 +20,17 @@
-    [EnabledIf='sharedPreferencesForWebProcess(connection).gamepadsEnabled'] StoppedUsingGamepads() -> ()
-    [EnabledIf='sharedPreferencesForWebProcess(connection).gamepadsEnabled'] PlayGamepadEffect(unsigned gamepadIndex, String gamepadID, enum:uint8_t WebCore::GamepadHapticEffectType type, struct WebCore::GamepadEffectParameters parameters) -> (bool success)
-    [EnabledIf='sharedPreferencesForWebProcess(connection).gamepadsEnabled'] StopGamepadEffects(unsigned gamepadIndex, String gamepadID) -> ()
+    [EnabledBy=GamepadsEnabled] StartedUsingGamepads()
```
The code is adding a mapping from connection to web process for each message.

This use case is mapping gaming message gamepads to web processes that are allowed to send the gamepad messages.

I think both would be better solved by moving the messages to a dedicated message listener.

E.g. using what's called "capability based access control".
* The access would  checked when the message listener is created.  E.g. caller sends message "create gamepad manager id 1", and then when handling this message, the preference would be checked.)
* If the listener exists, it is a proof that the sender is allowed to do the operation 

This is on the grounds of the capability based access control is already used, as it's the more general method and needed everywhere. In other words, everywhere we need to split up the complex functionality to submodules (like audio destination managers, graphics contexts, etc. and not everything is added to a single massive interface such as WebProcessPool ). 


Additional bug seems to be that StartedUsingGamepads are added to WebProcessPool instead of WebProcess.
Current logic:
* WebProcess A: send a message to WebProcessPool
* UI: Look up WebProcessProxy of the sender based on IPC::Connection
* UI: Look up preferences of the WebProcessProxy

More natural logic:
* WebProcess A: send a message to WebProcessProxy
* UI: Look up preferences of the WebProcessProxy

Same thing for the gamepad logic: it's not good to send to WebProcessPool if the first thing the code does is to look up WebProcessProxy. That means that the message is intended to WebProcessProxy, and sending it somewhere else is losing information.

rniwa replied:
I've discussed that design but the gamepads state need to be shared across all web processes so we're keeping the current design.

rniwa committed:
Add the support for generating sharedPreferencesForWebProcess which takes connection
https://bugs.webkit.org/show_bug.cgi?id=277475

Reviewed by Chris Dumez.

This PR adds the support for SharedPreferencesNeedsConnection extended attribute on
an IPC message receiver which, if specified on a message receiver, will generate
sharedPreferencesForWebProcess calls with connection as an argument. Also changed
the delimiter for specifying multiple feature flags from "," to "&&".

In addition, this PR consolidates multiple calls to sharedPreferencesForWebProcess
in the generated didReceiveMessage function to one call at the beginning of
the function to avoid the overhead of calling the function multiple times.

Finally, use this new capability in WebProcessPool.messages.in.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(generate_enabled_by):
(generate_runtime_enablement):
(async_message_statement):
(sync_message_statement):
(generate_enabled_by_for_receiver):
(generate_message_handler):
* Source/WebKit/Scripts/webkit/model.py:
(MessageReceiver.__init__):
(generate_global_model):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::messageArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledBy.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
(WebKit::TestWithEnabledBy::didReceiveMessage):
(IPC::jsValueForDecodedMessage<MessageName::TestWithEnabledBy_MultiConditionallyEnabled>):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessages.h:
(Messages::TestWithEnabledBy::MultiConditionallyEnabled::name):
(Messages::TestWithEnabledBy::MultiConditionallyEnabled::arguments):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessageReceiver.cpp:
(WebKit::TestWithEnabledIf::didReceiveMessage):
* Source/WebKit/UIProcess/WebProcessPool.messages.in:

Canonical link: https://commits.webkit.org/281977@main
